platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/io/ParallelOutputStream.kt

```diff
@@ -0,0 +1,248 @@
+     */
+    fun of(createOutputStream: () -> OutputStream): OutputStream {
+        val buffers = ByteBufferPool()
+        val ready = ConcurrentLinkedQueue<ByteBuffer>()
+        val writer = thread(name = "CC writer", isDaemon = true, priority = Thread.NORM_PRIORITY) {
+            try {
+                createOutputStream().use { outputStream ->
+                    val outputChannel = Channels.newChannel(outputStream)
+                    while (true) {
+                        val buffer = ready.poll()
+                        if (buffer == null) {
+                            // give the producer another chance
+                            Thread.yield()
+                            continue
+                        }
+                        if (!buffer.hasRemaining()) {
+                            /** producer is signaling end of stream
+                             * see [QueuedOutputStream.close]
+                             **/
+                            break
+                        }
+                        try {
+                            outputChannel.write(buffer)
+                        } finally {
+                            // always return the buffer
+                            buffers.put(buffer)
+                        }
+                    }
+                }
+            } catch (e: Exception) {
+                buffers.fail(e)
+            } finally {
+                // in case of failure, this releases some memory until the
+                // producer realizes there was a failure
+                ready.clear()
+            }
+            logger.debug {
+                "${javaClass.name} writer ${Thread.currentThread()} finished."
+            }
+        }
+        return QueuedOutputStream(buffers, ready) {
+            writer.join()
+        }
+    }
+}
+
+
+/**
+ * An [OutputStream] implementation that writes to buffers taken from a [ByteBufferPool]
+ * and posts them to the given [ready] queue when they are full.
+ */
+private
+class QueuedOutputStream(
+    private val buffers: ByteBufferPool,
+    private val ready: Queue<ByteBuffer>,
+    private val onClose: () -> Unit,
+) : OutputStream() {
+
+    private
+    var buffer = buffers.take()
+
+    override fun write(b: ByteArray, off: Int, len: Int) {
+        writeByteArray(b, off, len)
+    }
+
+    override fun write(b: Int) {
+        buffer.put(b.toByte())
+        maybeFlush()
+    }
+
+    override fun close() {
+        // send remaining data
+        if (buffer.position() > 0) {
+            sendBuffer()
+            takeNextBuffer()
+        }
+        // send a last empty buffer to signal the end
+        sendBuffer()
+        onClose()
+        buffers.rethrowFailureIfAny()
+        super.close()
+    }
+
+    private
+    tailrec fun writeByteArray(b: ByteArray, off: Int, len: Int) {
+        if (len <= 0) {
+            return
+        }
+        val remaining = buffer.remaining()
+        if (remaining > len) {
```

Comment:
❓ Should it be `>=`?

Reply:
Indeed, `>=` will avoid an iteration step.